internal/comm/catchup_handler.go

```diff
@@ -0,0 +1,137 @@
+// SPDX-License-Identifier: Apache-2.0
+
+package comm
+
+import (
+	"fmt"
+	"mime/multipart"
+	"net/http"
+
+	"github.com/IBM-Blockchain/bcdb-server/internal/httputils"
+	"github.com/IBM-Blockchain/bcdb-server/pkg/logger"
+	"github.com/IBM-Blockchain/bcdb-server/pkg/types"
+	"github.com/golang/protobuf/proto"
+	"github.com/gorilla/mux"
+)
+
+const (
+	BCDBPeerEndpoint = "/bcdb-peer/"
+	GetBlocksPath    = BCDBPeerEndpoint + "blocks"
+	GetHeightPath    = BCDBPeerEndpoint + "height"
+)
+
+//go:generate counterfeiter -o mocks/ledger_reader.go --fake-name LedgerReader . LedgerReader
+
+type LedgerReader interface {
+	Height() (uint64, error)
+	Get(blockNumber uint64) (*types.Block, error)
+}
+
+type catchupHandler struct {
+	router       *mux.Router
+	lg           *logger.SugarLogger
+	ledgerReader LedgerReader
+}
+
+func NewCatchupHandler(lg *logger.SugarLogger, ledgerReader LedgerReader) *catchupHandler {
+	h := &catchupHandler{
+		router:       mux.NewRouter(),
+		lg:           lg,
+		ledgerReader: ledgerReader,
+	}
+
+	h.router.HandleFunc(GetBlocksPath, h.blocksRequest).Methods(http.MethodGet).Headers("Accept", "multipart/form-data").Queries("start", "{startId:[0-9]+}", "end", "{endId:[0-9]+}")
+	h.router.HandleFunc(GetHeightPath, h.heightRequest).Methods(http.MethodGet)
+
+	return h
+}
+
+func (h *catchupHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	h.lg.Debugf("request: %s", r.URL)
+	h.router.ServeHTTP(w, r)
+}
+
+func (h *catchupHandler) blocksRequest(response http.ResponseWriter, request *http.Request) {
+	params := mux.Vars(request)
+	startBlockNum, endBlockNum, err := httputils.GetStartAndEndBlockNum(params)
+	if err != nil {
+		httputils.SendHTTPResponse(response, http.StatusBadRequest, &types.HttpResponseErr{ErrMsg: err.Error()})
+		return
+	}
+
+	height, err := h.ledgerReader.Height()
+	if err != nil {
+		httputils.SendHTTPResponse(response, http.StatusInternalServerError, &types.HttpResponseErr{ErrMsg: err.Error()})
+		return
+	}
+
+	if startBlockNum < 1 {
+		httputils.SendHTTPResponse(response, http.StatusBadRequest, &types.HttpResponseErr{fmt.Sprintf("requested startId [%d] must be greater than 0", startBlockNum)})
+		return
+	}
+
+	if startBlockNum > height {
+		httputils.SendHTTPResponse(response, http.StatusBadRequest, &types.HttpResponseErr{fmt.Sprintf("requested startId [%d] is out of range, height is [%d]", startBlockNum, height)})
+		return
+	}
+
+	var blocks [][]byte
+	for i := startBlockNum; i <= endBlockNum; i++ {
+		if i > height {
+			break
+		}
+		block, err := h.ledgerReader.Get(i)
+		if err != nil {
+			httputils.SendHTTPResponse(response, http.StatusInternalServerError, &types.HttpResponseErr{ErrMsg: err.Error()})
+			return
+
+		}
+		blockBytes, err := proto.Marshal(block)
+		if err != nil {
+			httputils.SendHTTPResponse(response, http.StatusInternalServerError, &types.HttpResponseErr{ErrMsg: err.Error()})
+			return
+
+		}
+		blocks = append(blocks, blockBytes)
```

Comment:
Not related to this PR but we need to add a `GetMarshaledBlock()` in the block store to avoid unmarshaling at block store and marshalling here.

Reply:
Agreed. Opened issue https://github.com/IBM-Blockchain/bcdb-server/issues/170